### PR TITLE
[TPU] For XLA Strategy, added function arg to control `broadcast_master_param()`

### PR DESCRIPTION
**Pull request by gkroiz**
## What does this PR do?
Adds boolean `sync_module_states` to `XLAStrategy` to control whether or not to call `broadcast_master_params()`. Broadcasting master parameters is not always needed. For example, when initializing random weights, if the seed is defined and the same on all devices, then broadcast_master_params() is not needed.

The motivation for this change is that `broadcast_master_params()` adds additional time to training and if not needed should be skipped.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/strategies/xla.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def __init__(
         parallel_devices: Optional[List[torch.device]] = None,
         checkpoint_io: Optional[CheckpointIO] = None,
         precision: Optional[Precision] = None,
+        broadcast_master_params: bool = True,
```
Comment (carmocca) [inline src/lightning/fabric/strategies/xla.py:49]: Since lightning has the `seed_everything` utility, we could check where [`PL_GLOBAL_SEED`](https://github.com/Lightning-AI/lightning/blob/9eedf7ae4b5cd398931af14ca7eac0605ef740ee/src/lightning/fabric/utilities/seed.py#L54) is available in the environment and that will tell us if we need to call `broadcast_master_params` and avoid this flag. What do you think?

Comment (gkroiz) [reply src/lightning/fabric/strategies/xla.py:49]: I had this thought too but have one concern. What happens if `seed_everything` is defined on each device with a different seed? For example: [Lit-Llama](https://github.com/Lightning-AI/lit-llama/blob/main/train_shakespeare.py#LL49C18-L49C18) defines the seed based on the device's rank `fabric.seed_everything(1337 + fabric.global_rank)`. In this case, would we need to call `broadcast_master_params`? If so, checking whether `PL_GLOBAL_SEED` exists in the environment would not be sufficient.

Comment (carmocca) [reply src/lightning/fabric/strategies/xla.py:49]: Good point. Normally we recommend using the same seed across ranks if the DistributedSampler is used, but that script doesn't use dataloaders at all.
cc @awaelchli

How much overhead does this add? Because ordinary DDP always does this

Comment (awaelchli) [reply src/lightning/fabric/strategies/xla.py:49]: @gkroiz @carmocca Yes correct, it really depends on the use case. Lit-llama defines it this way to make sure we don't end up with the same data samples on each rank. One could have also used the torch utils DataLoader and DistributedSampler pattern as Carlos said. Two different ways, same purpose. For simplicity there, it was chosen not to introduce the dataloader pattern.

Regarding broadcasting the parameters. DDP does it by default yes, while FSDP exposes it as an argument `sync_module_states` (https://pytorch.org/docs/stable/fsdp.html?highlight=fsdp#module-torch.distributed.fsdp) that is False by default. I would say it is fine to leave this decision up to the strategy. I would rather not tie it to the global seed setting, since this is quite a silent behavior that the user wouldn't be able to override.

**Comment by gkroiz** [comment]
> We should add a test like this one: https://github.com/Lightning-AI/lightning/pull/17370/files#diff-664848c0cba1d81c012e047af0b592829a44870a55e75c24884569e81e3e62beR41

I added a test for the Fabric XLAStrategy.

File: tests/tests_fabric/strategies/test_xla.py [hunk 2]
```diff
@@ -148,3 +151,32 @@ def tpu_all_gather_fn(strategy):
...
+            assert gathered[0] == t
+    else:
+        with pytest.raises(AssertionError):
+            for t in gathered:
+                assert gathered[0] == t
+
+
+@RunIf(tpu=True)
+@pytest.mark.parametrize("broadcast_master_params", [True, False])
+@mock.patch.dict(os.environ, {"PJRT_DEVICE": "TPU"}, clear=True)
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_xla.py:172]: This isn't necessary. Our CI sets it: https://github.com/Lightning-AI/lightning/actions/runs/4877179869/jobs/8701590389?pr=17522#step:7:66

```suggestion
```

File: tests/tests_fabric/strategies/test_xla.py [hunk 3]
```diff
@@ -148,3 +151,32 @@ def tpu_all_gather_fn(strategy):
 def test_tpu_all_gather():
     """Test the all_gather operation on TPU."""
     xla_launch(tpu_all_gather_fn)
+
+
+def tpu_broadcast_master_params_fn(broadcast_master_params, strategy):
+    seed_everything(random.seed())
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_xla.py:157]: This takes an integer

```suggestion
    seed_everything(123)
```

**Review by carmocca — APPROVED** [review]
LGTM! Last bit is to add a CHANGELOG entry for both Fabric and PyTorch

File: tests/tests_fabric/strategies/test_xla.py (hunk 3 above)
Comment (gkroiz) [reply tests/tests_fabric/strategies/test_xla.py:157]: Whoops you are right, this should not have `random.seed()`. However, `123` as an input will break the current test since the test currently tests when the broadcast_master_params is set to `True` and `False`. Leaving the input as blank will set the seeds to be different for each device. I will change this when I get back.

**Review by Borda — APPROVED** [review]

File: src/lightning/fabric/strategies/xla.py [hunk 4]
```diff
@@ -57,6 +58,7 @@ def __init__(
         self._checkpoint_io: Optional[CheckpointIO]
         self._backward_sync_control = None  # XLA synchronizes gradients in the optimizer.step() call
         self._launched = False
+        self._broadcast_master_params = broadcast_master_params
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/xla.py:61]: Name suggestion to avoid "master" in the name:
- broadcast_main_params
- broadcast_rank_0_params
- broadcast_params

Comment (carmocca) [reply src/lightning/fabric/strategies/xla.py:61]: If we are not going with the same name as xla's, then we could also consider `sync_module_states` to match FSDP

Comment (gkroiz) [reply src/lightning/fabric/strategies/xla.py]: Changed to sync_module_states
